cpp/src/arrow/memory_pool.cc

```diff
@@ -603,14 +643,109 @@ class BaseMemoryPoolImpl : public MemoryPool {
     stats_.UpdateAllocatedBytes(-size);
   }
 
-  void ReleaseUnused() override { Allocator::ReleaseUnused(); }
+ protected:
+  virtual Status AllocateImmutableZeros(int64_t size, uint8_t** out) {
+#ifdef USE_MMAP_FOR_IMMUTABLE_ZEROS
+    if (size > 0) {
+      *out = static_cast<uint8_t*>(mmap(
+          nullptr, size, PROT_READ, MAP_PRIVATE | MAP_ANONYMOUS | MAP_NORESERVE, -1, 0));
+      if (*out == MAP_FAILED) {
+        auto err = errno;
+        return Status::OutOfMemory("Failed to allocate zero buffer of size ", size, ": ",
+                                   strerror(err));
+      }
+      return Status::OK();
+    }
+#endif
+    RETURN_NOT_OK(Allocate(size, out));
+    std::memset(*out, 0, size);
+    return Status::OK();
+  }
+
+  void FreeImmutableZeros(uint8_t* buffer, int64_t size) override {
+#ifdef USE_MMAP_FOR_IMMUTABLE_ZEROS
+    if (size > 0) {
+      munmap(buffer, size);
+      return;
+    }
+#endif
+    Free(buffer, size);
+  }
+
+ public:
+  Result<std::shared_ptr<Buffer>> GetImmutableZeros(int64_t size) override {
+    // Thread-safely get the current largest buffer of zeros.
```

Comment:
This algorithm seems quite complicated and costly (in terms of inter-thread synchronization).
Instead, how about maintaining lazily-allocated power-of-two-sized entries (perhaps starting with 4096 to avoid having a ton of tiny size classes)? Something like:
```c++

static constexpr int kMinPower2Size = 12;  // 4096 bytes
static constexpr int kNumAllocClasses = 64 - kMinPower2Size;
// Addresses of allocated zeros, for each size class
std::array<std::atomic<uintptr_t>, kNumAllocClasses> allocated_;
std::mutex allocation_mutex_;

Result<const uint8_t*> GetImmutableZeros(int64_t size) override {
  const auto size_class = std::max(0, bit_util::Log2(size) - kMinPower2Size);
  auto addr = allocated_[size_class].load();
  if (ARROW_PREDICT_TRUE(addr != nullptr)) {
    // Fast path: already allocated
    return reinterpret_cast<const uint8_t*>(addr);
  }
  // Slow path: allocate if not already done
  std::lock_guard<std::mutex> lock(allocation_mutex_);
  auto addr = allocated_[size_class].load();
  if (addr == nullptr) {
    const int64_t alloc_size = static_cast<int64_t>(1) << (size_class + kMinPower2Size);
    ARROW_ASSIGN_OR_RAISE(const uint8_t* data, AllocateImmutableZeros(size));
    allocated_[size_class] = addr = reinterpret_cast<uintptr_t>(data);
  }
  return reinterpret_cast<const uint8_t*>(addr);
}
```

Reply:
Compared to my algorithm:

 - `+` In terms of thread primitives, your fast path only involves a memory fence, whereas mine involves a mutex. I'm not sure if it's even needed, though: it isn't if you can safely make a copy of a `shared_ptr` while another thread may be updating the contained pointer. That feels like a true statement, at least on x86, but I couldn't figure it out for sure from the C++ docs. If I can remove it, my fast path is just a `shared_ptr` copy (so, an atomic increment), a null check, and a size check, which I'm pretty sure is the fastest way to do it that implements reference counting for deallocation.
 - `+` Your version doesn't allocate unnecessarily small buffers.
 - `+` Your version is more readable, especially seeing how unnecessarily cryptic I wrote the reallocation logic.
 - `-` Your version has no way to free buffers, so I would argue that it leaks memory. Granted, it's upper-bounded by a bit less than 2x the next larger power-of-two of the largest buffer allocated, so it won't grow without bound. By comparison however, my version will release smaller buffers when they are no longer used, and will free its cache when `ReleaseUnused()` is called and there are no other users. I also considered a version where the cache is a `weak_ptr`, in which case the `ReleaseUnused()` would not be needed, but decided against it mostly because `ReleaseUnused()` already existed.
 - `-` Nit, but your version will allocate small buffers regardless of whether a larger buffer is already available, whereas my version will return the largest buffer allocated thus far, and will automatically free previously allocated smaller buffers when all their users go out of scope.
 - `-` Also kind of a nit, but rounding up to power-of-two-sized buffers means that you might throw an out of memory error even if almost half of the requested memory isn't needed. My algorithm will back off and allocate only as much as is needed if the 2 * previous size allocation fails.

An inability to free something, especially if that something is large, feels like bad news to me, so I'm hesitant to just copy your version in and call it a day. But if nothing else, I'll add a lower bound for allocation size and try to rewrite the allocation algorithm to be less cryptic tomorrow.

Reply:
> An inability to free something, especially if that something is large, feels like bad news to me, so I'm hesitant to just copy your version in and call it a day.

That is true. However, it shouldn't be a concern if we can ensure that the pages don't actually allocate physical memory (or almost none of it, such as `/dev/zero`).

Reply:
> * I'm not sure if it's even needed, though: it isn't if you can safely make a copy of a `shared_ptr` while another thread may be updating the contained pointer.

You can safely update the object _pointed to_. However, changing the _pointer_ requires use of dedicated atomic access functions: https://en.cppreference.com/w/cpp/memory/shared_ptr/atomic

Reply:
> That is true. However, it shouldn't be a concern if we can ensure that the pages don't actually allocate physical memory (or almost none of it, such as `/dev/zero`).

I don't think we can in general, unfortunately. I would have no idea how to do it on something as ubiquitous as Windows (or if it can be done at all), and I'm sure that in general there are more exotic operating systems and architectures that simply can't do it. Also, for 32-bit systems/builds (if Arrow supports those) virtual memory is also in relatively short supply.

> However, changing the pointer requires use of dedicated atomic access functions: [...]

Ah, great, those functions are exactly what I had missed!

I improved my allocation algorithm accordingly here: e628688

Reply:
Unfortunately [the atomic operations don't exist in libstdc++ 4.9 and down](https://gcc.gnu.org/bugzilla/show_bug.cgi?id=57250), as still used in CI via RTools 35. So I guess in that case, my algorithm has to resort to a mutex after all. f67f75c

Reply:
Ah, sorry. You can just use `arrow/util/atomic_shared_ptr.h` :-)

Reply:
Ah, yeah, that's certainly a lot cleaner. dee486a